noir-projects/noir-protocol-circuits/crates/types/src/abis/accumulated_data/private_accumulated_data_builder.nr

```diff
@@ -0,0 +1,189 @@
+
+    private_call_stack: BoundedVec<CallRequest, MAX_PRIVATE_CALL_STACK_LENGTH_PER_TX>,
+    public_call_stack: BoundedVec<CallRequest, MAX_PUBLIC_CALL_STACK_LENGTH_PER_TX>,
+}
+
+impl PrivateAccumulatedDataBuilder {
+    pub fn finish(self) -> PrivateAccumulatedData {
+        PrivateAccumulatedData {
+            new_note_hashes: self.new_note_hashes.storage,
+            new_nullifiers: self.new_nullifiers.storage,
+            new_l2_to_l1_msgs: self.new_l2_to_l1_msgs.storage,
+            encrypted_logs_hash: self.encrypted_logs_hash,
+            unencrypted_logs_hash: self.unencrypted_logs_hash,
+            encrypted_log_preimages_length: self.encrypted_log_preimages_length,
+            unencrypted_log_preimages_length: self.unencrypted_log_preimages_length,
+            private_call_stack: self.private_call_stack.storage,
+            public_call_stack: self.public_call_stack.storage
+        }
+    }
+
+    pub fn to_combined(self) -> CombinedAccumulatedData {
+        CombinedAccumulatedData {
+            new_note_hashes: self.new_note_hashes.storage,
+            new_nullifiers: self.new_nullifiers.storage,
+            new_l2_to_l1_msgs: self.new_l2_to_l1_msgs.storage,
+            encrypted_logs_hash: self.encrypted_logs_hash,
+            unencrypted_logs_hash: self.unencrypted_logs_hash,
+            encrypted_log_preimages_length: self.encrypted_log_preimages_length,
+            unencrypted_log_preimages_length: self.unencrypted_log_preimages_length,
+            public_data_update_requests: unsafe::zeroed()
+        }
+    }
+
+    pub fn split_to_public(
+        self,
+        min_revertible_side_effect_counter: u32
+    ) -> (PublicAccumulatedData, PublicAccumulatedData) {
+        let mut non_revertible_builder: PublicAccumulatedDataBuilder = unsafe::zeroed();
+        let mut revertible_builder: PublicAccumulatedDataBuilder = unsafe::zeroed();
+
+        for i in 0..MAX_NEW_NOTE_HASHES_PER_TX {
+            let note_hash = self.new_note_hashes.storage[i];
+            if note_hash.counter < min_revertible_side_effect_counter {
+                non_revertible_builder.new_note_hashes.push(note_hash);
+            } else {
+                revertible_builder.new_note_hashes.push(note_hash);
+            }
+        }
+        for i in 0..MAX_NEW_NULLIFIERS_PER_TX {
+            let nullifier = self.new_nullifiers.storage[i];
+            if nullifier.counter < min_revertible_side_effect_counter {
+                non_revertible_builder.new_nullifiers.push(nullifier);
+            } else {
+                revertible_builder.new_nullifiers.push(nullifier);
+            }
+        }
+
+        for i in 0..MAX_PUBLIC_CALL_STACK_LENGTH_PER_TX {
+            let call_stack_item = self.public_call_stack.storage[i];
+            if call_stack_item.start_side_effect_counter < min_revertible_side_effect_counter {
+                non_revertible_builder.public_call_stack.push(call_stack_item);
```

Comment:
Should we assert after this loop that we have exactly two public call stack items? since the code relies on that for needs_setup and needs_teardown

Reply:
It should be able to handle the case where there is >=1 public call stack item:
It should be that the public enqueued call that is last should be for teardown, and all other should be setup.